platform/engineering/request-for-change/2020-01-08-ssoe-removing-implicit-mvi-queries.md

```diff
@@ -0,0 +1,106 @@
+# RFC: Removing implicit MVI lookups
+
+- Date: _2020-01-08_
+
+
+## Background
+At present, when a user authenticates and is determined to be Level of assurance 3, we look up that user's VA profile in MVI (aka MPI, used interchangeably hereafter). The combination of the incoming attributes from the identity provider and the attributes from MVI make up the bulk of the logical user model that is used across vets-api.
+
+As part of the overall SSO consolidation work, we are integrating VA SSOe with the VA.gov sign-in process. We will maintain the three current credential options, but the SAML request/response cycle will be intercepted by SSOe before being returned to VA.gov. Along the way, SSOe can establish a session cookie for use by other sites. But more importantly, SSOe will provide a standardized SAML payload that includes attributes from the credential provider AND from MVI. 
+
+Since we will be receiving the bulk of the MVI attributes in the SAML payload, we want to determine whether we can and should do away with the separate MVI query that normally happens immediately following authentication.
+
+## Motivation
+Given that we will receive a SAML payload with most of the needed information from MVI already, doing a separate query to MVI is largely redundant, and adds latency to the sign-in process (though note we're likely swapping added latency during the SAML round-trip for the MVI query latency).
+
+
+
+## Design
+In general, MVI attributes are accessed via the `User` model, either as direct delegated methods on `User` (e.g. `user.icn`), or as nested delegated methods on `User.va_profile` (e.g. `user.profile.address). Both of these eventually delegate to the `mvi.profile` attribute on `user`, which is an instance of `MviProfile`.  
+
+We can instead instantiate a MviProfile object by parsing the incoming SAML attributes and delegate to that directly from `User.va_profile`. Initially we'll want to do so selectively for SSOe authentication so we can support old and new authentication mechanisms during roll-out. 
```

Comment:
Is this possibly an opportunity to have all of the MVI profile information just live directly on the user model or I guess more accurately user_identity model? I think this would simplify things quite a bit.

Reply:
Yes, logically if it comes in via a SAML assertion it will just be attributes on user_identity, and then I would just change the  delegation of `user.va_profile` things to point to `identity`.